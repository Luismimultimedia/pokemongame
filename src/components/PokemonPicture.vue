<template>
    <div class="pokemon-container">
        <img
            v-if="!showPokemon"
            class="hidden-pokemon"
            :src="imgSrc"
            alt="pokemon"
        />
        <img v-else class="fade-in" :src="imgSrc" alt="pokemon" />
    </div>
</template>

<script>
export default {
    name: "PokemonPicture",
    props: {
        pokemonId: {
            default: 1,
            type: Number,
            required: true,
        },
        showPokemon: {
            default: false,
            type: Boolean,
            required: true,
        },
    },
    computed: {
        imgSrc() {
            return `https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/${this.pokemonId}.svg`;
        },
    },
};
</script>

<style scoped>
img {
    height: 200px;
    position: absolute;
    left: 0;
    right: 0;
    margin-left: auto;
    margin-right: auto;
    user-select: none;
    -moz-user-select: none;
    -ms-user-select: none;
    -webkit-user-drag: none;
    -webkit-user-select: none;
}

.hidden-pokemon {
    filter: brightness(0);
}

.pokemon-container {
    height: 200px;
}
</style>
