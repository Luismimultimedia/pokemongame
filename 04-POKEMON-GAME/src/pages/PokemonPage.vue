<template>
    <h1 v-if="!pokemon" class="fade-in">Cargando la información...</h1>
    <div v-else>
        <h1>¿Quién es este pokémon?</h1>
        <PokemonPicture :pokemonId="pokemon?.id" :showPokemon="showPokemon" />
        <PokemonOption :pokemons="pokemonArr" @selection="checkAnswer" />
        <template v-if="showAnswer">
            <h3>{{ message }}</h3>
            <button @click="resetGame()">Reiniciar</button>
        </template>
    </div>
</template>

<script>
import PokemonPicture from "@/components/PokemonPicture";
import PokemonOption from "@/components/PokemonOption";

import getPokemonsOptions from "@/helpers/getPokemonOptions";

export default {
    name: "PokemonPage",
    components: {
        PokemonPicture,
        PokemonOption,
    },
    data() {
        return {
            pokemonArr: [],
            pokemon: null,
            showPokemon: false,
            showAnswer: false,
            message: "",
        };
    },
    methods: {
        async mixPokemonArray() {
            this.pokemonArr = await getPokemonsOptions();
            const rnfInt = Math.floor(Math.random() * 4);
            this.pokemon = this.pokemonArr[rnfInt];
        },
        checkAnswer(data) {
            this.showPokemon = data.id === this.pokemon.id;
            this.showAnswer = true;
            this.message = this.showPokemon
                ? "Eres un maestro pokemon"
                : "Haz fallado";
        },
        resetGame() {
            this.pokemonArr = [];
            this.pokemon = null;
            this.showPokemon = false;
            this.showAnswer = false;
            this.message = "";
            this.mixPokemonArray();
        },
    },
    mounted() {
        this.mixPokemonArray();
    },
};
</script>
