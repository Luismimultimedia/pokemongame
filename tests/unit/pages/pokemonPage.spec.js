import { shallowMount, mount } from "@vue/test-utils";
import PokemonPage from "@/pages/PokemonPage"
import PokemonPicture from "@/components/PokemonPicture";
import PokemonOptions from "@/components/PokemonOption";

import { pokemons } from "../mocks/pokemons.mock";

describe('Pruebas unitarias en PokemonPage Page', () => {

    let wrapper

    beforeEach(() => {
        wrapper = shallowMount(PokemonPage)
    });

    test('Debe de hacer match con el snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot()
    });

    test('Debe de llamar el mixPokemonArray al montar', () => {
        const mixPokemonArraySpy = jest.spyOn(PokemonPage.methods, 'mixPokemonArray')
        const wrapper = shallowMount(PokemonPage)
        expect(mixPokemonArraySpy).toHaveBeenCalled()
    });

    test('Debe de hacer match con el snapshot cuando cargan los pokemons', () => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ""
                }
            }
        })

        expect(wrapper.html()).toMatchSnapshot()
    });

    test('Debe de mostrar los componentes de PokemonPicture y PokemonOptions', () => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ""
                }
            }
        })
        const pokemonPictureComponent = wrapper.findComponent(PokemonPicture)
        const pokemonOptionsComponent = wrapper.findComponent(PokemonOptions)
        const { pokemonId } = pokemonPictureComponent.props()
        const { pokemons: arrPokemon } = pokemonOptionsComponent.props()

        expect(pokemonPictureComponent.exists()).toBeTruthy()
        expect(pokemonOptionsComponent.exists()).toBeTruthy()
        expect(pokemonId).toBe(151)
        expect(arrPokemon).toStrictEqual(pokemons)
    });

    test('Pruebas con checkAnswer', async () => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ""
                }
            }
        })

        await wrapper.vm.checkAnswer({ id: 151 })

        expect(wrapper.find('h3').exists()).toBeTruthy()
        expect(wrapper.vm.showPokemon).toBeTruthy()
    });
});