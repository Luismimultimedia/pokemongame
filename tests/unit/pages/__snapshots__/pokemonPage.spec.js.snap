// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas unitarias en PokemonPage Page Debe de hacer match con el snapshot 1`] = `<h1 class="fade-in">Cargando la información...</h1>`;

exports[`Pruebas unitarias en PokemonPage Page Debe de hacer match con el snapshot cuando cargan los pokemons 1`] = `
<div>
  <h1>¿Quién es este pokémon?</h1>
  <pokemon-picture-stub pokemonid="151" showpokemon="false"></pokemon-picture-stub>
  <pokemon-option-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-option-stub>
  <!--v-if-->
</div>
`;
